{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.7.12","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"# Input data files are available in the read-only \"../input/\" directory\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","execution":{"iopub.status.busy":"2022-07-21T18:27:17.963034Z","iopub.execute_input":"2022-07-21T18:27:17.963445Z","iopub.status.idle":"2022-07-21T18:27:17.975556Z","shell.execute_reply.started":"2022-07-21T18:27:17.963413Z","shell.execute_reply":"2022-07-21T18:27:17.974525Z"},"trusted":true},"execution_count":2,"outputs":[]},{"cell_type":"markdown","source":"##  <font color='purple'>Analysis of Lung single cell RNA-seq data</font>\n\n\nAim of this project is to identify cell types and annotate the clusters of lung tissue sc RNA sequencing data. The data used in this analyis is a [patient adjacent non-tumour involved lung tissue](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3516673) sample produced by Laughney et al. (2020) via the 10× Genomics Chromium platform. \n\n\n**Data Analysis Outline**\n\n1. Load data into Anndata\n\n2. Preprocessing\n\n 2.1. Normalization\n\n 2.2. Feature selection\n\n 2.3. Quality control\n\n3. Feature Reduction and Visualization\n\n  3.1 Principal Component Analysis (PCA)\n\n4. Downstream Analysis\n\n 4.1. Clustering\n \n 4.2 Differential Expression Analysis\n \n 4.3. Marker Identification\n \n 4.4. Cluster Annotation using CellO classifier\n \n 4.5. Uniform Manifold Approximation and Projection (UMAP)\n \n 4.6. Conclusion\n \n \n **Background Information: scRNA‐seq workflow**\n\nSingle cell data generation has multiple steps single‐cell dissociation, single‐cell isolation, library construction, and sequencing (Ziegenhain et al, 2017). The input material is biological tissue sample and single-cell suspension is generated by tissue digestion (single-cell dissociation). Cells are isolated in different ways depending on experimental protocol to profile mRNA in each cell separately. Plate-based method isolates cell in wells on a plate and droplet-based methods capture cells in microfluidic droplet. Some of the challenges are multiples cells being captured (doublets), non-viables cells or no cell being captures in wells/ droplets.\n\nEach well/droplet contains chemicals to break cell membrane and perform library construction. In library contruction, intracellular mRNA is captured, reverse-transcribed to cDNA molecules and amplified (to increase the probability of being measured). During isolation, mRNA from each cell can be lebelled with well/droplet cellular barcode or unique molecular identifier (UMI). UMIs helps to distinguish between amplified copies of the same mRNA molecule and reads from separate mRNA molecules transcribed from the same gene (Luecken MD et al, Theis FJ et al, 2019).\n\nThe libraries generated are grouped together for sequencing. Read data from sequencing undergoes quality control, grouping based on barcodes and alignment. Read data is demultiplexed to generate count data in UMI-based protocols (Luecken MD et al, Theis FJ et al, 2019).","metadata":{}},{"cell_type":"code","source":"# Install and load required packages\n!pip install scanpy\n!pip install scvi\n!pip3 install leidenalg\n!sudo apt-get install graphviz libgraphviz-dev graphviz-dev pkg-config -y\n!pip install pygraphviz\n!pip install leidenalg cello-classify","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"import scanpy as sc\nimport pandas as pd\nimport numpy as np\nfrom matplotlib.pyplot import rc_context\n\nimport scvi\nimport os\nfrom anndata import AnnData\nimport cello","metadata":{"execution":{"iopub.status.busy":"2022-07-21T18:29:03.867487Z","iopub.execute_input":"2022-07-21T18:29:03.868292Z","iopub.status.idle":"2022-07-21T18:29:08.427516Z","shell.execute_reply.started":"2022-07-21T18:29:03.868236Z","shell.execute_reply":"2022-07-21T18:29:08.426347Z"},"trusted":true},"execution_count":4,"outputs":[]},{"cell_type":"code","source":"sc.set_figure_params(dpi=100, color_map = 'viridis_r')\nsc.settings.verbosity = 1\nsc.logging.print_header()","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## 1. Load data into Anndata","metadata":{}},{"cell_type":"code","source":"data_path =\"../input/gsm3516673-lung-tissue/GSM3516673_MSK_LX682_NORMAL_dense.csv\"","metadata":{"execution":{"iopub.status.busy":"2022-07-21T18:29:08.429132Z","iopub.execute_input":"2022-07-21T18:29:08.429801Z","iopub.status.idle":"2022-07-21T18:29:08.434896Z","shell.execute_reply.started":"2022-07-21T18:29:08.429766Z","shell.execute_reply":"2022-07-21T18:29:08.433785Z"},"trusted":true},"execution_count":5,"outputs":[]},{"cell_type":"code","source":"# load the expression matrix into a Pandas df\ndf = pd.read_csv(data_path,index_col=0)\ndf.head()","metadata":{"execution":{"iopub.status.busy":"2022-07-21T18:29:08.437545Z","iopub.execute_input":"2022-07-21T18:29:08.437904Z","iopub.status.idle":"2022-07-21T18:29:43.387974Z","shell.execute_reply.started":"2022-07-21T18:29:08.437871Z","shell.execute_reply":"2022-07-21T18:29:43.386942Z"},"trusted":true},"execution_count":6,"outputs":[]},{"cell_type":"markdown","source":"Anndata is cell by gene expression matrix with row as cells and columns as genes and used by Scanpy library.","metadata":{}},{"cell_type":"code","source":"# Load the expression matrix using Pandas and Scanpy\nadata = AnnData(df)\nadata","metadata":{"execution":{"iopub.status.busy":"2022-07-21T18:29:43.389351Z","iopub.execute_input":"2022-07-21T18:29:43.389697Z","iopub.status.idle":"2022-07-21T18:29:43.616030Z","shell.execute_reply.started":"2022-07-21T18:29:43.389668Z","shell.execute_reply":"2022-07-21T18:29:43.614629Z"},"trusted":true},"execution_count":7,"outputs":[]},{"cell_type":"markdown","source":"## 2. Preprocessing","metadata":{}},{"cell_type":"code","source":"# Basic filtering\nsc.pp.filter_cells(adata, min_genes=200) #  filter cells with fewer 200 genes detected\nsc.pp.filter_genes(adata, min_cells=3) # filtered out genes that are detected in less than 3 cells","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### 2.1. Normalization\n\nAs single-cell consists of wide-ranging cells with varying size and counts, it is important to scale the count data. Commonly used method is counts per million” (CPM), which normalizes count data using size factor proportional to the count depth per cell. \n\nNormalization can be performed based on the type of scRNA-seq techniques: full‐length and 3′ enrichment methods (Svensson et al, 2017; Ziegenhain et al, 2017, and Luecken MD et al, Theis FJ et al, 2019). The gene length normalization is particularly useful for full-length data and not for 3′ enrichment (e.g. Patel et al, 2014; Kowalczyk et al, 2015; Soneson & Robinson, 2018, and Luecken MD et al, Theis FJ et al, 2019)\n\n\nAfter normalization, data is log(x+1)‐transformed as it will be required for next steps such as batch correction or differential expression analysis, which assume normal distribution and requires log transformation. Three uses of log transformation: changes in expression is meaured by log fold changes, reduction of mean-variance relationship in single-cell data and skewness (Brennecke et al, 2013; Finak et al, 2015; Ritchie et al, 2015; Johnson et al, 2006; Buttner et al, 2019).","metadata":{}},{"cell_type":"code","source":"# Normalize expression data into units of log transcripts per million\nsc.pp.normalize_total(adata, target_sum=1e6)\n\n# Logarithmize the data:\nsc.pp.log1p(adata)","metadata":{"execution":{"iopub.status.busy":"2022-07-21T18:29:45.183554Z","iopub.execute_input":"2022-07-21T18:29:45.183977Z","iopub.status.idle":"2022-07-21T18:29:45.993284Z","shell.execute_reply.started":"2022-07-21T18:29:45.183944Z","shell.execute_reply":"2022-07-21T18:29:45.991993Z"},"trusted":true},"execution_count":10,"outputs":[]},{"cell_type":"markdown","source":"### 2.2 Feature Sellection: Selecting Highly variables genes\n\nSelection of Higly variables genes (n_top_genes) is used for downstream analysis. Selection of high number of gene decreases the separation between different cell types. Selecting few genes will not result in clear distinction of granular cell types (Bernstein et al. 2021).","metadata":{}},{"cell_type":"code","source":"# Annotate the 10000 most highly expressed genes for performing clustering\nsc.pp.highly_variable_genes(adata, n_top_genes=10000)\n# sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)","metadata":{"execution":{"iopub.status.busy":"2022-07-21T18:29:45.995056Z","iopub.execute_input":"2022-07-21T18:29:45.995868Z","iopub.status.idle":"2022-07-21T18:29:48.258118Z","shell.execute_reply.started":"2022-07-21T18:29:45.995819Z","shell.execute_reply":"2022-07-21T18:29:48.257148Z"},"trusted":true},"execution_count":11,"outputs":[]},{"cell_type":"code","source":"#Show those genes that yield the highest fraction of counts in each single cell, across all cells.\nsc.pl.highest_expr_genes(adata, n_top=20,)","metadata":{"execution":{"iopub.status.busy":"2022-07-21T18:29:48.262893Z","iopub.execute_input":"2022-07-21T18:29:48.263263Z","iopub.status.idle":"2022-07-21T18:29:49.334002Z","shell.execute_reply.started":"2022-07-21T18:29:48.263230Z","shell.execute_reply":"2022-07-21T18:29:49.332737Z"},"trusted":true},"execution_count":12,"outputs":[]},{"cell_type":"markdown","source":"\n### 2.3 Quality control\n\nThere are three QC covariates (Ilicic et al, 2016; Griffiths et al, 2018, and Luecken MD et al, Theis FJ et al, 2019)\n\n1. Count depth (the number of counts per barcode)\n2. Number of genes per barcode\n3. Fraction of counts from mitochondrial genes per barcode \n\n\nA violin plot is used for visualizing QC covariates. This helps in detection of outliers which correpond to dying cells, broken cell membrane and doublets. \nFor example, cytoplasmic mRNA leakout through broken membrane is indicated by low count depth, few genes and high fraction of mitochondrial counts.\nConversely, high count depth and high gene detection indicated doublets (DoubletDecon: preprint: DePasquale et al, 2018 and Luecken MD et al, Theis FJ et al, 2019).","metadata":{}},{"cell_type":"code","source":"adata.var['mt'] = adata.var_names.str.startswith('MT-') # annotate the group of mitochondrial genes as 'mt'\nsc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)","metadata":{"execution":{"iopub.status.busy":"2022-07-21T19:32:56.322462Z","iopub.execute_input":"2022-07-21T19:32:56.322921Z","iopub.status.idle":"2022-07-21T19:32:56.923248Z","shell.execute_reply.started":"2022-07-21T19:32:56.322888Z","shell.execute_reply":"2022-07-21T19:32:56.921955Z"},"trusted":true},"execution_count":33,"outputs":[]},{"cell_type":"code","source":"sc.pl.violin(adata, ['n_genes_by_counts', 'total_counts', 'pct_counts_mt'],\n             jitter=0.4, multi_panel=True)","metadata":{"execution":{"iopub.status.busy":"2022-07-21T19:33:01.254911Z","iopub.execute_input":"2022-07-21T19:33:01.255359Z","iopub.status.idle":"2022-07-21T19:33:01.943422Z","shell.execute_reply.started":"2022-07-21T19:33:01.255321Z","shell.execute_reply":"2022-07-21T19:33:01.942530Z"},"trusted":true},"execution_count":34,"outputs":[]},{"cell_type":"code","source":"# Plot mitochondrial genes expressed and total counts\nsc.pl.scatter(adata, x='total_counts', y='pct_counts_mt')\nsc.pl.scatter(adata, x='total_counts', y='n_genes_by_counts')","metadata":{"execution":{"iopub.status.busy":"2022-07-21T19:33:05.907051Z","iopub.execute_input":"2022-07-21T19:33:05.908339Z","iopub.status.idle":"2022-07-21T19:33:06.386900Z","shell.execute_reply.started":"2022-07-21T19:33:05.908286Z","shell.execute_reply":"2022-07-21T19:33:06.385985Z"},"trusted":true},"execution_count":35,"outputs":[]},{"cell_type":"code","source":"# Filtering to remove cells that have too many mitochondrial genes expressed or too many total counts\nadata = adata[adata.obs.n_genes_by_counts < 2500, :]\nadata = adata[adata.obs.pct_counts_mt < 5, :]","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"#adata = adata[:, adata.var.highly_variable]","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Normalized vs Non-normalized mean expression of genes\nsc.pl.highly_variable_genes(adata)\n","metadata":{"execution":{"iopub.status.busy":"2022-07-21T19:33:19.970227Z","iopub.execute_input":"2022-07-21T19:33:19.970988Z","iopub.status.idle":"2022-07-21T19:33:20.604603Z","shell.execute_reply.started":"2022-07-21T19:33:19.970930Z","shell.execute_reply":"2022-07-21T19:33:20.603326Z"},"trusted":true},"execution_count":36,"outputs":[]},{"cell_type":"markdown","source":"Even after normalization, a count depth effect can remain in the data. Regressing out count depth effects will increase performance of trajectory inference algorithms which find the transition between cells (Luecken MD et al, Theis FJ et al 2019)\n","metadata":{}},{"cell_type":"code","source":"# Regress out effects of total counts per cell and the percentage of mitochondrial genes expressed\nsc.pp.regress_out(adata, ['total_counts', 'pct_counts_mt'])","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### 3.1  Principal Component Analysis\n\nReduces dimensionality of the data. n_comps computes number of principal components. ","metadata":{}},{"cell_type":"code","source":"# principal components=5\nsc.pp.pca(adata, n_comps=50, use_highly_variable=True)","metadata":{"execution":{"iopub.status.busy":"2022-07-21T18:30:00.107627Z","iopub.execute_input":"2022-07-21T18:30:00.108287Z","iopub.status.idle":"2022-07-21T18:30:09.217438Z","shell.execute_reply.started":"2022-07-21T18:30:00.108225Z","shell.execute_reply":"2022-07-21T18:30:09.216117Z"},"trusted":true},"execution_count":14,"outputs":[]},{"cell_type":"code","source":"sc.pl.pca(adata)\n#sc.pl.pca_variance_ratio(adata, log=True)","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### 4.1 Clustering ","metadata":{}},{"cell_type":"code","source":"# Compute nearest-neighbors graph by finding the nearest 10 cells to each cell using the Euclidean distance \nsc.pp.neighbors(adata, n_neighbors=15)\nsc.tl.leiden(adata, resolution=2.0) # Resolution: High values for this parameter result in smaller, more granular clusters","metadata":{"execution":{"iopub.status.busy":"2022-07-21T19:33:42.696985Z","iopub.execute_input":"2022-07-21T19:33:42.697464Z","iopub.status.idle":"2022-07-21T19:33:43.923230Z","shell.execute_reply.started":"2022-07-21T19:33:42.697428Z","shell.execute_reply":"2022-07-21T19:33:43.921893Z"},"trusted":true},"execution_count":38,"outputs":[]},{"cell_type":"markdown","source":"### 4.2 Differential Expression Analysis\n\nMarker genes are identified by computing differential expression between groups: Cells in one cluster versus rest of the clusters. The genes that are upregulated or top-ranked genes in a cluster are its marker genes. Most commonly used methods are t-test and Wilcoxon rank-sum (Mann-Whitney-U). Using reference databases cells in the clusters can be identified and annotated.\n\nIn this analysis, we are using CellO, a Python package for annotating cell types in single-cell RNA-seq data (Matthew N. Bernstein et al, Colin N. Dewey et al 2021) to identify the cell types and annotate the clusters.","metadata":{}},{"cell_type":"code","source":"# computing differential expression between groups using t-test\nsc.tl.rank_genes_groups(adata,'leiden', method='t-test')","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### 4.3 Marker Identification","metadata":{}},{"cell_type":"code","source":"# Ranking the upregulated genes in each cluster\nsc.pl.rank_genes_groups(adata, n_genes=25, sharey=False) \n#sc.settings.set_figure_params(dpi=160)","metadata":{"execution":{"iopub.status.busy":"2022-07-21T18:44:01.731564Z","iopub.execute_input":"2022-07-21T18:44:01.731970Z","iopub.status.idle":"2022-07-21T18:44:07.985872Z","shell.execute_reply.started":"2022-07-21T18:44:01.731927Z","shell.execute_reply":"2022-07-21T18:44:07.984596Z"},"trusted":true},"execution_count":20,"outputs":[]},{"cell_type":"code","source":"# Show the 25 top ranked genes per cluster 0, 1, …, 7 in a dataframe.\npd.DataFrame(adata.uns['rank_genes_groups']['names']).head(20)\n","metadata":{"execution":{"iopub.status.busy":"2022-07-21T19:36:44.506342Z","iopub.execute_input":"2022-07-21T19:36:44.506742Z","iopub.status.idle":"2022-07-21T19:36:44.641662Z","shell.execute_reply.started":"2022-07-21T19:36:44.506710Z","shell.execute_reply":"2022-07-21T19:36:44.640168Z"},"trusted":true},"execution_count":42,"outputs":[]},{"cell_type":"code","source":"# Get a table with the scores and groups.\nresult = adata.uns['rank_genes_groups']\ngroups = result['names'].dtype.names\npd.DataFrame(\n    {group + '_' + key[:1]: result[key][group]\n    for group in groups for key in ['names', 'pvals']}).head()","metadata":{"execution":{"iopub.status.busy":"2022-07-21T18:44:08.116523Z","iopub.execute_input":"2022-07-21T18:44:08.116862Z","iopub.status.idle":"2022-07-21T18:44:08.179410Z","shell.execute_reply.started":"2022-07-21T18:44:08.116832Z","shell.execute_reply":"2022-07-21T18:44:08.178240Z"},"trusted":true},"execution_count":22,"outputs":[]},{"cell_type":"code","source":"# Marker_genes for cluster 0\nmarker_genes = [\"CYP27A1\", \"HNMT\", \"FTH1\",\"MCEMP1\",\"ALOX5\",\"MRC1\",\"MSR1\",\"S100A13\",\"CXCL16\",\"PTMS\",\"INHBA\", \"SLC11A1\",\"PDLIM1\",\"SPI1\",\"TREM1\",\"MS4A7\",\"NUPR1\",\"OLR1\",\"LINC01272\",\"LGALS3BP\"]","metadata":{"execution":{"iopub.status.busy":"2022-07-21T19:38:52.322667Z","iopub.execute_input":"2022-07-21T19:38:52.323758Z","iopub.status.idle":"2022-07-21T19:38:52.331693Z","shell.execute_reply.started":"2022-07-21T19:38:52.323713Z","shell.execute_reply":"2022-07-21T19:38:52.330370Z"},"trusted":true},"execution_count":43,"outputs":[]},{"cell_type":"code","source":"# mean expression of marker_genes from cluster 0 across all clusters\nsc.pl.dotplot(adata, marker_genes, groupby='leiden');","metadata":{"execution":{"iopub.status.busy":"2022-07-21T19:38:54.070896Z","iopub.execute_input":"2022-07-21T19:38:54.071343Z","iopub.status.idle":"2022-07-21T19:38:54.767941Z","shell.execute_reply.started":"2022-07-21T19:38:54.071297Z","shell.execute_reply":"2022-07-21T19:38:54.767106Z"},"trusted":true},"execution_count":44,"outputs":[]},{"cell_type":"markdown","source":"### 4.4 Cluster Annotation using CellO classifier","metadata":{}},{"cell_type":"code","source":"cello_resource_loc = os.getcwd()","metadata":{"execution":{"iopub.status.busy":"2022-07-21T18:30:56.956726Z","iopub.execute_input":"2022-07-21T18:30:56.957255Z","iopub.status.idle":"2022-07-21T18:30:56.970639Z","shell.execute_reply.started":"2022-07-21T18:30:56.957216Z","shell.execute_reply":"2022-07-21T18:30:56.969424Z"},"trusted":true},"execution_count":18,"outputs":[]},{"cell_type":"code","source":"model_prefix = \"GSM3516673_MSK_LX682_NORMAL\" # <-- The trained model will be stored in a file called GSM3516666_LX682_NORMAL.model.dill \n\ncello.scanpy_cello(\n    adata, \n    clust_key='leiden',\n    rsrc_loc=cello_resource_loc, \n    out_prefix=model_prefix,\n    log_dir=os.getcwd()\n)","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"After pre-training model with the genes in the target expression matrix, then we can run CellO below without training a new model in the step below.","metadata":{}},{"cell_type":"code","source":"\n#cello.scanpy_cello(\n    #adata,\n    #clust_key='leiden',\n    #rsrc_loc=cello_resource_loc,\n    #model_file=f'{model_prefix}.model.dill'\n#)","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### 4.5 UMAP \n\nCells identified with Cello classifier are visualized using UMAP","metadata":{}},{"cell_type":"code","source":"sc.tl.umap(adata)","metadata":{"execution":{"iopub.status.busy":"2022-07-21T19:27:49.568278Z","iopub.execute_input":"2022-07-21T19:27:49.569606Z","iopub.status.idle":"2022-07-21T19:27:57.935264Z","shell.execute_reply.started":"2022-07-21T19:27:49.569552Z","shell.execute_reply":"2022-07-21T19:27:57.934023Z"},"trusted":true},"execution_count":28,"outputs":[]},{"cell_type":"code","source":"fig = sc.pl.umap(adata, color='leiden', title='Clusters', return_fig=True)","metadata":{"execution":{"iopub.status.busy":"2022-07-21T19:28:07.775090Z","iopub.execute_input":"2022-07-21T19:28:07.775974Z","iopub.status.idle":"2022-07-21T19:28:08.682979Z","shell.execute_reply.started":"2022-07-21T19:28:07.775936Z","shell.execute_reply":"2022-07-21T19:28:08.682092Z"},"trusted":true},"execution_count":29,"outputs":[]},{"cell_type":"code","source":"# Create UMAP plot with cells colored by most-specific predicted cell type\nfig = sc.pl.umap(adata, color='Most specific cell type', return_fig=True)","metadata":{"execution":{"iopub.status.busy":"2022-07-21T19:28:12.693681Z","iopub.execute_input":"2022-07-21T19:28:12.694720Z","iopub.status.idle":"2022-07-21T19:28:13.224055Z","shell.execute_reply.started":"2022-07-21T19:28:12.694672Z","shell.execute_reply":"2022-07-21T19:28:13.222846Z"},"trusted":true},"execution_count":30,"outputs":[]},{"cell_type":"code","source":"# UMAP showing the probability of being T cells\nfig = sc.pl.umap(adata, color='T cell (probability)', vmin=0.0, vmax=1.0, return_fig=True)","metadata":{"execution":{"iopub.status.busy":"2022-07-21T19:28:25.825337Z","iopub.execute_input":"2022-07-21T19:28:25.825728Z","iopub.status.idle":"2022-07-21T19:28:26.055623Z","shell.execute_reply.started":"2022-07-21T19:28:25.825698Z","shell.execute_reply":"2022-07-21T19:28:26.054244Z"},"trusted":true},"execution_count":31,"outputs":[]},{"cell_type":"code","source":"# UMAP showing the cells being classified as being T cells (Orange = T cells)\nfig = sc.pl.umap(adata, color='T cell (binary)', return_fig=True)","metadata":{"execution":{"iopub.status.busy":"2022-07-21T19:28:30.109094Z","iopub.execute_input":"2022-07-21T19:28:30.109494Z","iopub.status.idle":"2022-07-21T19:28:30.318422Z","shell.execute_reply.started":"2022-07-21T19:28:30.109461Z","shell.execute_reply":"2022-07-21T19:28:30.317456Z"},"trusted":true},"execution_count":32,"outputs":[]},{"cell_type":"markdown","source":"### 4.6 Conclusion\n\nFrom the cell identification and cluster annotation analysis, the scRNA sequencing data of lung tissue has predominantly following cell types\n\n* CD1c-positive myeloid dendritic cell\n\n* CD4-positive, alpha-beta memory T cell\n\n* CD14-positive, CD16-positive monocyte\n\n* alveiolar macrophage\n\n* dendritic cell,human \n\n* effector memory CD4-positive, alpha-beta T cell\n\n* endo-epithelial cell\n\n* endothelial cell\n\n* epithelial cell of upper respiratory tract\n\n* memory B cell\n\n* myeloid cell\n\n* myeloid leukocyte\n\n* natural killer cell\n\n* plasmocytoid dentritic cell\n\n* type II pneumocyte\n\n\n","metadata":{}},{"cell_type":"markdown","source":"### References\n\n1. Luecken MD, Theis FJ. Current best practices in single-cell RNA-seq analysis: a tutorial. Mol Syst Biol. 2019 Jun 19;15(6):e8746. doi: 10.15252/msb.20188746. PMID: 31217225; PMCID: PMC6582955.\n\n2. Matthew N. Bernstein, Colin N. Dewey,Annotating cell types in human single-cell RNA-seq data with CellO,STAR Protocols,\nVolume 2, Issue 3,2021,100705,ISSN 2666-1667,https://doi.org/10.1016/j.xpro.2021.100705 (https://www.sciencedirect.com/science/article/pii/S2666166721004123)\n\n3. Ziegenhain C, Vieth B, Parekh S, Reinius B, Guillaumet-Adkins A, Smets M, Leonhardt H, Heyn H, Hellmann I, Enard W (2017) Comparative analysis of single-cell RNA sequencing methods. Mol Cell 65: 631–643.e4\n\n4. Svensson V, Natarajan KN, Ly L-H, Miragaia RJ, Labalette C, Macaulay IC, Cvejic A, Teichmann SA (2017) Power analysis of single-cell RNA-sequencing experiments. Nat Methods 14: 381–387\n\n5. Patel AP, Tirosh I, Trombetta JJ, Shalek AK, Gillespie SM, Wakimoto H, Cahill DP, Nahed BV, Curry WT, Martuza RL et al (2014) Single-cell RNA-seq highlights intratumoral heterogeneity in primary glioblastoma. Science 344: 1396–1401\n\n6. Kowalczyk MS, Tirosh I, Heckl D, Rao TN, Dixit A, Haas BJ, Schneider RK, Wagers AJ, Ebert BL, Regev A (2015) Single-cell RNA-seq reveals changes in cell cycle and differentiation programs upon aging of hematopoietic stem cells. Genome Res 25: 1860–1872 \n\n7. Soneson C, Robinson MD (2018) Bias, robustness and scalability in single-cell differential expression analysis. Nat Methods 15: 255–261\n\n8. Brennecke P, Anders S, Kim JK, Kołodziejczyk AA, Zhang X, Proserpio V, Baying B, Benes V, Teichmann SA, Marioni JC et al (2013) Accounting for technical noise in single-cell RNA-seq experiments. Nat Methods 10: 1093–1095\n\n9. Finak G, McDavid A, Yajima M, Deng J, Gersuk V, Shalek AK, Slichter CK, Miller HW, Juliana McElrath M, Prlic M et al (2015) MAST: a flexible statistical framework for assessing transcriptional changes and characterizing heterogeneity in single-cell RNA sequencing data.\n\n10. Ritchie ME, Phipson B, Wu D, Hu Y, Law CW, Shi W, Smyth GK (2015) limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Res 43: e47\n\n11. Johnson WE, Li C, Rabinovic A (2006) Adjusting batch effects in microarray expression data using empirical Bayes methods.\n\n12. Buttner M, Miao Z, Wolf A, Teichmann SA, Theis FJ (2019) A test metric for assessing single-cell RNA-seq batch correction. Nat Methods 16: 43–49\n\n13. Ilicic T, Kim JK, Kolodziejczyk AA, Bagger FO, McCarthy DJ, Marioni JC, Teichmann SA (2016) Classification of low quality cells from single-cell RNA-seq data. Genome Biol 17: 29\n\n14. Griffiths JA, Scialdone A, Marioni JC (2018) Using single-cell genomics to understand developmental processes and cell fate decisions. Mol Syst Biol 14: e8046","metadata":{}},{"cell_type":"markdown","source":"**Appendix**\n\nPandas : Pandas is an open source library providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language\n\nMatplotlib : Matplotlib is a comprehensive library for creating static, animated, and interactive visualizations in Python\n\nSCVI: Probabilistic models for single-cell omics data\n\nScanpy: Scanpy is an open sourcy library for analyzing single-cell gene expression data built jointly with anndata in python\n\nAnndata: Python package for handling annotated data matrices in memory and on disk, positioned between pandas and xarray\n\nCellO: CellO is a Python package for annotating cell types in single-cell RNA-seq data\n","metadata":{}}]}